File: Blind 75/String/Valid Parentheses.java
class Solution {
    public boolean isValid(String s) {
        // Time O(n) and Space O(n) worst case
        Stack<Character> stack = new Stack<>();
        for (char ch : s.toCharArray()) {
            if (ch == '(' || ch == '{' || ch == '[') { // opening parentheses gets added
                stack.add(ch);
            } else if (!stack.isEmpty() && ch == ')' && stack.peek() == '('
                || !stack.isEmpty() && ch == '}' && stack.peek() == '{'
                || !stack.isEmpty() && ch == ']' && stack.peek() == '[') { // stack being empty means there's no opening parentheses
                stack.pop();
            } else {
                return false; // stack is empty and there is a closing parentheses
            }
        }
        return stack.isEmpty(); // could have a trailing open parentheses
    }
}

File: Python Blind 75/String/Valid Parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        # Time: O(n), Space: O(n) in case all the chars are opening brackets stored in stack
        stack = []
        closeToOpen = {")" : "(", "}" : "{", "]" : "[",}

        for ch in s: # Loops through each char in string s
            if ch in closeToOpen: # checks if the current char is one of the keys in the dict
                if stack and stack[-1] == closeToOpen[ch]: # checking if stack is not empty and the stack peek is the corresponding open parenthesis
                    stack.pop()
                else:
                    return False
            else:
                stack.append(ch)

        return True if not stack else False